<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
  <!-- muscle is true for muscle hand or false for motor hand -->
  <!-- hand_type can be either hand_e (default) or hand_e_plus -->
  <!-- prefix can be anything but usually is "r_" or  "l_" for right and left hands distinction -->
  <!-- cable_mesh is set to true when cable collision should be taken into account -->
  <xacro:macro name="dark_grey" >
    <material name="dark_grey">
      <color rgba="0.21961 0.21961 0.21961 1.0" />
    </material>
  </xacro:macro>
  <xacro:macro name="light_grey" >
    <material name="light_grey">
      <color rgba="0.90000 0.90000 0.90000 1.0" />
    </material>
  </xacro:macro>
  <xacro:macro name="forearm" params="muscle hand_type:='hand_e' prefix cable_mesh:='false'">
    <link name="${prefix}forearm">
      <inertial>
        <origin xyz="0 0 0.09" rpy="0 0 0" />
        <mass value="3.0" />
        <inertia ixx="0.0138" ixy="0.0" ixz="0.0" iyy="0.0138"
        iyz="0.0" izz="0.00744" />
      </inertial>
      <xacro:if value="${muscle}">
        <visual>
          <origin xyz="0 0 0" rpy="0 0 0" />
          <geometry name="${prefix}forearm_visual">
            <mesh filename="package://sr_description/meshes/hand/forearm_muscle.dae" scale="1.0 1.0 1.0" />
          </geometry>
          <material name="Grey" />
        </visual>
      </xacro:if>
      <xacro:unless value="${muscle}">
        <xacro:if value="${hand_type == 'hand_e'}">
          <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry name="${prefix}forearm_visual">
              <mesh filename="package://sr_description/meshes/hand/forearm.dae" scale="0.001 0.001 0.001" />
            </geometry>
            <material name="light_grey" />
          </visual>
        </xacro:if>
        <xacro:if value="${hand_type == 'hand_e_plus'}">
          <visual>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry name="${prefix}forearm_visual">
              <mesh filename="package://sr_description/meshes/hand/forearm_visual_plus.dae" scale="0.001 0.001 0.001" />
            </geometry>
            <material name="light_grey" />
          </visual>
        </xacro:if>
      </xacro:unless>
      <collision>
        <xacro:if value="${muscle}">
          <origin xyz="0 0 0.185" rpy="0 0 0" />
          <geometry name="${prefix}forearm_collision">
            <cylinder radius="0.075" length="0.37" />
          </geometry>
          <xacro:dark_grey/>
        </xacro:if>
        <xacro:unless value="${muscle}">
          <xacro:if value="${hand_type == 'hand_e'}">
            <origin xyz="0 0 0.092" rpy="0 0 0" />
            <geometry name="${prefix}forearm_collision">
              <cylinder radius="0.07" length="0.184"/>
            </geometry>
            <xacro:dark_grey/>
          </xacro:if>
          <xacro:if value="${hand_type == 'hand_e_plus'}">
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry name="${prefix}forearm_collision">
              <mesh filename="package://sr_description/meshes/hand/forearm_collision_plus.stl" scale="0.001 0.001 0.001" />
            </geometry>
            <xacro:dark_grey/>
          </xacro:if>
        </xacro:unless>
      </collision>
      <!-- wrist mount -->
      <collision>
        <xacro:if value="${muscle}">
          <origin xyz="0 0 0.395" rpy="0 0 0" />
          <geometry>
            <box size="0.03 0.04 0.06" />
          </geometry>
        </xacro:if>
        <xacro:unless value="${muscle}">
          <origin xyz="0 -0.01 0.181" rpy="0 0.78 0" />
          <geometry>
            <box size="0.07 0.07 0.07" />
          </geometry>
        </xacro:unless>
      </collision>
      <xacro:if value="${cable_mesh}">
        <visual>
          <origin rpy="0 -1.57 -1.57" xyz="0.0 0.067 0.0045"/>
          <geometry name="${prefix}cable_extr_visualisation">
            <mesh filename="package://sr_description/meshes/hand/cable_collision_mesh.stl" scale="1 1 1" />
          </geometry>
        </visual>
        <collision>
          <origin rpy="0 -1.57 -1.57" xyz="0.0 0.067 0.0045"/>
          <geometry name="${prefix}cable_extr_collision">
            <mesh filename="package://sr_description/meshes/hand/cable_collision_mesh.stl" scale="1 1 1" />
          </geometry>
        </collision>
      </xacro:if>
    </link>
    <gazebo reference="${prefix}forearm">
      <selfCollide>false</selfCollide>
    </gazebo>
  </xacro:macro>
</robot>
